include/encodings/Encodings.hpp

```diff
@@ -0,0 +1,86 @@
+#ifndef Encodings_hpp
+#define Encodings_hpp
+
+#include <iostream>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <utility>
+#include <functional>
+#include <z3++.h>
+#include <cmath>
+#include <map>
+#include <limits>
+
+#include "utils.hpp"
+
+using namespace z3;
+
+struct NestedVar
+{
+	explicit NestedVar(unsigned long varID) : varID(varID), list() {};
+	NestedVar(unsigned long varID, std::vector<NestedVar>  list) : varID(varID), list(std::move(list))
+	{
+	}
+	unsigned long varID = std::numeric_limits<unsigned long>::max();
+	std::vector<NestedVar> list;
+};
+
+struct WeightedVar
+{
+	WeightedVar(unsigned long varID, int weight) : varID(varID), weight(weight) {}
+	unsigned long varID = std::numeric_limits<unsigned long>::max();
+	int weight = 0;
+};
+inline
+bool operator<(const WeightedVar& rhs, const WeightedVar& lhs) {
+	return rhs.weight < lhs.weight;
+}
+inline
+bool operator==(const WeightedVar& rhs, const WeightedVar& lhs) {
+	return rhs.weight == lhs.weight && rhs.varID == lhs.varID;
+}
+
+struct SavedLit
+{
+	SavedLit() :type(-1), varID(0) {}
+	SavedLit(int type, unsigned long varID) : type(type), varID(varID) {}
+	int type = -1;
```

Comment:
Wouldn't it be more explanatory if `type` was not merely an `int` but an inner enum class `Type` with its possible values? Something like:
```suggestion
        enum class Type { neg, pos } // whatever the possible types are
	Type type = Type::pos; // or whatever the default shall be
```